gocourse03/main.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	Chip       RodentType = "Chip"
+	Dale       RodentType = "Dale"
+	Cheaser    RodentType = "Cheaser"
+	Hackwrench RodentType = "Hackwrench"
+
+	NorthSector  Sector = "North"
+	SouthSector  Sector = "South"
+	WestSector   Sector = "West"
+	EastSector   Sector = "East"
+	CenterSector Sector = "Center"
+)
+
+type RodentType string
+
+type Sector string
+
+func main() {
+	var movements []Movement
+
+	rodents := []Rodent{
+		{ID: 1, RodentType: Chip, FromTo: [2]Sector{WestSector, WestSector}},
+		{ID: 2, RodentType: Dale, FromTo: [2]Sector{EastSector, EastSector}},
+		{ID: 3, RodentType: Cheaser, FromTo: [2]Sector{SouthSector, SouthSector}},
+		{ID: 4, RodentType: Hackwrench, FromTo: [2]Sector{NorthSector, NorthSector}},
+	}
+
+	westRodents := []Rodent{rodents[0]}
+	eastRodents := []Rodent{rodents[1]}
+	southRodents := []Rodent{rodents[2]}
+	northRodents := []Rodent{rodents[3]}
+
+	param := RodentParametr{
+		Rodents:      &rodents,
+		WestRodents:  &westRodents,
+		EastRodents:  &eastRodents,
+		SouthRodents: &southRodents,
+		NorthRodents: &northRodents,
+		Movements:    &movements,
+	}
+
+	PrintLocations(&rodents, &movements)
+
+	moveThroughCenter(1, time.Now().Add(10), WestSector, EastSector, param)
+	moveThroughCenter(2, time.Now().Add(11), EastSector, NorthSector, param)
+	moveThroughCenter(2, time.Now().Add(12), NorthSector, SouthSector, param)
+	moveThroughCenter(3, time.Now().Add(13), SouthSector, WestSector, param)
+
+	PrintLocations(&rodents, &movements)
+}
+
+type Rodent struct {
+	ID         int
+	RodentType RodentType
+	FromTo     [2]Sector
+}
+
+type Movement struct {
+	MoveTime time.Time
+	FromTo   [2]Sector
+}
+
+func moveThroughCenter(
+	id int,
+	moveTime time.Time,
+	from Sector,
+	to Sector,
+	param RodentParametr,
+) {
+	rIndex := Index(*param.Rodents, id)
+	(*param.Rodents)[rIndex].FromTo[0] = from
+	(*param.Rodents)[rIndex].FromTo[1] = to
+
+	dst := FindSectorSlice(to, param.WestRodents, param.EastRodents, param.NorthRodents, param.SouthRodents)
+	*dst = append(*dst, (*param.Rodents)[rIndex])
+
+	src := FindSectorSlice(from, param.WestRodents, param.EastRodents, param.NorthRodents, param.SouthRodents)
+	index := Index(*src, id)
+	*src = append((*src)[:index], (*src)[index+1:]...)
+
+	*param.Movements = append(*param.Movements, Movement{MoveTime: moveTime, FromTo: [2]Sector{from, to}})
+}
+
+func FindSectorSlice(
+	name Sector,
+	westRodents *[]Rodent,
+	eastRodents *[]Rodent,
+	northRodents *[]Rodent,
+	southRodents *[]Rodent) *[]Rodent {
+	switch name {
+	case WestSector:
+		return westRodents
+	case EastSector:
+		return eastRodents
+	case NorthSector:
+		return northRodents
+	case SouthSector:
+		return southRodents
+	default:
+		return nil
+	}
+}
+
+func Index(rodentsSlice []Rodent, id int) int {
```

Comment:
`Slice` зайвий суфікс:
```suggestion
func Index(rodents []Rodent, id int) int {
```